### PR TITLE
Refactor grading bar configuration

### PR DESCRIPTION
**Pull request by marcospri**
Motivated to accommodate:

- https://github.com/hypothesis/lms/issues/5621

The launch view was dealing with the differences around grading in Canvas vs the rest of the LMSes.

The "instructor toolbar" was mixed with grading / editing & canvas. 

Made the editing vs grading aspects of the toolbar more explicit and moved the canvas way of doing things around grading part of the misc plugin.

Better solution than:

- https://github.com/hypothesis/lms/pull/5469

## Non canvas

- Delete grading infos

```docker compose exec postgres psql -U postgres -c "truncate lis_result_sourcedid cascade"```

- As `aunltd.S1` launch [PDF document in URL](https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2424/View?ou=6782)

- As `HypothesisEng.Teacher` launch [PDF document in URL](https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2424/View?ou=6782)

The tool bar will show up with the student there

## Canvas
eng+canvasstudent@hypothes.is

- As `eng+canvasstudent@hypothes.is` launch  [localhost (make devdata) HTML Assignment](https://hypothesis.instructure.com/courses/125/assignments/873)

Make an annotation.

- As eng+canvasteacher@hypothes.is launch  [localhost (make devdata) HTML Assignment in speed grader mode ](https://hypothesis.instructure.com/courses/125/gradebook/speed_grader?assignment_id=873&student_id=35)

Switching between students focuses their annotations (go to "Hypothesis 101 student" to see the annotation made earlier.

### CODE REVIEW

File: lms/product/canvas/_plugin/grading.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        # For students in Canvas with grades to submit we need to enable
+        # Speedgrader settings for gradable assignments
+        # `lis_result_sourcedid` associates a specific user with an
+        # assignment.
+        if (
+            assignment.is_gradable
+            and lti_user.is_learner
+            and lti_params.get("lis_result_sourcedid")
+        ):
+            js_config.add_canvas_speedgrader_settings(assignment.document_url)
```
Comment (jon-betts) [inline lms/product/canvas/_plugin/grading.py:21]: Should these settings be here? If this is Canvas specific, and this is the place that calls it, seems like the time?

File: lms/product/canvas/_plugin/grading.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+        # We add a `focused_user` query param to the SpeedGrader LTI launch
+        # URLs we submit to Canvas for each student when the student
+        # launches an assignment. Later, Canvas uses these URLs to launch
+        # us when a teacher grades the assignment in SpeedGrader.
+        if focused_user := request.params.get("focused_user"):
+            js_config.set_focused_user(focused_user)
```
Comment (jon-betts) [inline lms/product/canvas/_plugin/grading.py:28]: I think this might also be specific to Canvas. Also is this specific to learners?

File: lms/product/canvas/_plugin/grading.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @staticmethod
+    def configure_grading_for_launch(request, js_config, assignment):
+        """
+        Configure grading during a launch.
+
+        In canvas we use Speed Grader, we configure the frontend to use it.
+        """
+        lti_params = request.lti_params
+        lti_user = request.lti_user
+
```
Comment (jon-betts) [inline lms/product/canvas/_plugin/grading.py:11]: I'm not sure this is getting us much. Maybe inline these guys?

File: lms/views/lti/basic_launch.py [hunk 4]
```diff
@@ -30,15 +30,19 @@ class BasicLaunchViews:
...
         self.assignment_service: AssignmentService = request.find_service(
             name="assignment"
         )
         self._guid = self.request.lti_params.get("tool_consumer_instance_guid")
         self._resource_link_id = self.request.lti_params.get("resource_link_id")
 
         self.request.lti_user.application_instance.check_guid_aligns(self._guid)
+        self.request.find_service(name="application_instance").update_from_lti_params(
+            self.request.lti_user.application_instance, self.request.lti_params
+        )
```
Comment (jon-betts) [inline lms/views/lti/basic_launch.py:44]: I like these being next to each other

File: lms/product/canvas/_plugin/grading.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+class CanvasGradingPlugin:
+    @staticmethod
+    def configure_grading_for_launch(request, js_config, assignment):
```
Comment (jon-betts) [inline lms/product/canvas/_plugin/grading.py:3]: Some type defs here could be nice

File: tests/unit/lms/product/canvas/_plugin/grading_test.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        )
+
+        if focused_user:
+            js_config.set_focused_user.assert_called_once_with(focused_user)
+        else:
+            js_config.set_focused_user.assert_not_called()
+
+    def test_factory(self, pyramid_request):
+        plugin = CanvasGradingPlugin.factory(sentinel.context, pyramid_request)
+        assert isinstance(plugin, CanvasGradingPlugin)
```
Comment (jon-betts) [inline tests/unit/lms/product/canvas/_plugin/grading_test.py:70]: I thought we were in Java land for a while here! What does the test factory create?

File: lms/product/plugin/grading.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    def configure_grading_for_launch(request, js_config, assignment):
+        """Configure grading during a launch."""
+        if request.lti_user.is_instructor:
+            # For instructors, display the toolbar
+            js_config.enable_instructor_toolbar(enable_grading=assignment.is_gradable)
+        else:
+            # Create or update a record of LIS result data for a student launch
+            # We'll query these rows to populate the student dropdown in the
+            # instructor toolbar
+            request.find_service(name="grading_info").upsert_from_request(request)
```
Comment (jon-betts) [inline lms/product/plugin/grading.py:12]: This method seems to do quite a few different possible things:

 * Show the toolbar with grading support
 * Show the toolbar for editing the assignment **<<< Nothing to do with grading at all?**
 * Store grading info for the user

So as far as the name goes: "configure_**grading**\_for_**launch**":

 * Grading applies 2/3 cases
 * Launch applies 3/3 cases

This makes me wonder if this is more of a launch plugin than a grading plugin? It's also only called from the launch view.

Some different potential ways of skinning this cat:

* A `LaunchPlugin` with `show_instructor_toolbar` and `configure_grading_for_student`
   * Canvas would do nothing for `show_instructor_toolbar` and put everything in `configure_grading_for_student`
   * We could also move the test for `is_instructor` outside of the plugins
* A `LaunchPlugin` with a more generic "pre-launch-hook" point

I think I prefer the more structured one though, and I think the commonality of checking for the student or not is correct to keep outside (at least as far as the examples we have so far.

I think if we had a `LaunchPlugin` that some of the other misc items could move there too.

**Review by jon-betts — COMMENTED** [review]
I like separating out the Canvas stuff and it simplifies the launch a lot, but I think there are some more things we can and should do here.

 * I think we can go further with splitting out the cases. I think there is a distinct instructor action and learner action. At the moment they are merged and we then test inside the plugins to separate them again.
 * My gut feel is this is a "LaunchPlugin" not a "GradingPlugin" and that some other items from Misc could go in there too (doesn't have to be this PR for the move, but the name feels right)
 * We've moved calls to `js_config` into the Canvas plugin but not moved the Canvas specific code it calls from `js_config`. I think we should as this would mean the plugin extracts non standard code out of the general code.
 * Also `js_config` is a mess and any help it can get would be good
 * Is `set_focused` user learner specific? It's not in the code but the comment suggests it should be

File: tests/unit/lms/product/canvas/_plugin/grading_test.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+import pytest
+
+from lms.product.canvas._plugin.grading import CanvasGradingPlugin
+from lms.resources._js_config import JSConfig
+from tests import factories
+
+
+class TestGradingPlugin:
+    @pytest.mark.usefixtures("user_is_learner", "with_student_grading_id")
```
[the remark this answers is not in the document]
Comment (marcospri) [reply tests/unit/lms/product/canvas/_plugin/grading_test.py]: Followed a similar approach

File: lms/product/plugin/grading.py (hunk 7 above)
Comment (marcospri) [reply lms/product/plugin/grading.py]: Split the toolbar options in two, editing and grading. Quite like the explicitly of this now :+1:

File: lms/product/canvas/_plugin/misc.py [hunk 9]
```diff
@@ -8,6 +8,34 @@
...
+            js_config.add_canvas_speedgrader_settings(assignment.document_url)
+
+        # We add a `focused_user` query param to the LTI launch URL
+        # we send to canvas with student submissions.
+        # Later, when Canvas uses these URLs to launch
+        # us when a teacher grades the assignment in SpeedGrader we read that value and
+        # filter the annotations on the client to only show the ones by that student.
+        if focused_user := request.params.get("focused_user"):
+            js_config.set_focused_user(focused_user)
+
```
Comment (marcospri) [inline lms/product/canvas/_plugin/misc.py:38]: Copied verbatim from the view. The goal is to move non-standard code (ie Canvas only) outside of the main flow.

File: lms/product/canvas/product.py [hunk 10]
```diff
@@ -24,3 +24,8 @@ class Canvas(Product):
     )
 
     settings_key = "canvas"
+
+    use_toolbar_grading = False
```
Comment (marcospri) [inline lms/product/canvas/product.py:28]: Canvas doesn't use either of these but they are orthogonal.

File: lms/product/product.py [hunk 11]
```diff
@@ -68,6 +68,11 @@ class Product:
     settings_key: Optional[str] = None
     """Key in the ai.settings dictionary that holds the product specific settings"""
 
+    use_toolbar_grading = True
```
Comment (marcospri) [inline lms/product/product.py:71]: The default is to use both

File: lms/resources/_js_config/__init__.py [hunk 12]
```diff
@@ -290,70 +290,32 @@ def add_deep_linking_api(self):
         self._config.setdefault("filePicker", {})
         self._config["filePicker"]["deepLinkingAPI"] = config
 
-    def enable_instructor_toolbar(self, enable_editing=True, enable_grading=False):
-        """
-        Enable the toolbar with controls for instructors in LMS assignments.
+    def enable_toolbar_editing(self):
```
Comment (marcospri) [inline lms/resources/_js_config/__init__.py:293]: Split the toolbar configuration in two, editing and grading.

File: lms/resources/_js_config/__init__.py [hunk 13]
```diff
@@ -290,70 +290,32 @@ def add_deep_linking_api(self):
...
-                        # We are using the value from the request instead of the one stored in GradingInfo.
-                        # This allows us to still read and submit grades when something in the LMS changes.
-                        # For example in LTI version upgrades, the endpoint is likely to change as we move from
-                        # LTI 1.1 basic outcomes API to LTI1.3's Assignment and Grade Services.
-                        # Also when the install's domain is updated all the records in the DB will be outdated.
-                        "LISOutcomeServiceUrl": self._request.lti_params[
-                            "lis_outcome_service_url"
-                        ],
-                    }
-                )
```
Comment (marcospri) [inline lms/resources/_js_config/__init__.py:343]: This is too much non-config stuff for `js_config` moved out.

File: lms/services/grading_info.py [hunk 14]
```diff
@@ -20,11 +22,17 @@ def __init__(self, _context, request):
         self._db = request.db
         self._authority = request.registry.settings["h_authority"]
 
-    def get_by_assignment(self, application_instance, context_id, resource_link_id):
+    def get_students_for_grading(
```
Comment (marcospri) [inline lms/services/grading_info.py:25]: We don't use `GradingInfos` outside the context of the "grading bar".

Moved setting all of that up here.

I reckon this could be improved by making this is the "ToolbarGrading" service and moving a few more bits here. I'll leave that for later.

File: lms/services/grading_info.py [hunk 15]
```diff
@@ -33,12 +41,47 @@ def get_by_assignment(self, application_instance, context_id, resource_link_id):
...
         )
+        students = []
+
+        for grading_info in grading_infos:
+            h_user = HUser(
+                username=grading_info.h_username,
+                display_name=grading_info.h_display_name,
+            )
+
+            lis_result_sourced_id = grading_info.lis_result_sourcedid
```
Comment (marcospri) [inline lms/services/grading_info.py:59]: All of this was part of JSConfig before.

File: lms/services/grading_info.py [hunk 16]
```diff
@@ -55,7 +98,7 @@ def upsert_from_request(self, request):
             # We're missing something we need in the request.
             # This can happen if the user is not a student, or if the needed
             # LIS data is not present on the request.
-            return
+            return None
```
Comment (marcospri) [inline lms/services/grading_info.py:101]: Return something in all branches.

File: lms/views/lti/basic_launch.py [hunk 17]
```diff
@@ -152,8 +156,41 @@ def _show_document(self, assignment):
             assignment, groupings=[self.course]
         )
 
+        if self.request.product.use_toolbar_editing:
```
Comment (marcospri) [inline lms/views/lti/basic_launch.py:159]: All of these toolbar stuff is here together

File: lms/views/lti/basic_launch.py [hunk 18]
```diff
@@ -152,8 +156,41 @@ def _show_document(self, assignment):
...
+            self.context.js_config.enable_toolbar_editing()
+
+        if self.request.product.use_toolbar_grading and assignment.is_gradable:
```
Comment (marcospri) [inline lms/views/lti/basic_launch.py:162]: This could be moved to another service or maybe just a private method her.

The main goal of this PR is to have all the grading bar bits together to make a change. Once that other change lands I'll move everything in one go.

File: tests/unit/conftest.py [hunk 19]
```diff
@@ -295,16 +296,12 @@ def lti_registration():
...
 @pytest.fixture
-def lti_v13_application_instance(db_session, lti_registration):
-    application_instance = factories.ApplicationInstance(
-        developer_key="TEST_DEVELOPER_KEY",
-        provisioning=True,
-        deployment_id="TEST_DEPLOYMENT_ID",
-        lti_registration=lti_registration,
-        settings=ApplicationSettings({}),
-    )
+def lti_v13_application_instance(db_session, application_instance, lti_registration):
```
Comment (marcospri) [inline tests/unit/conftest.py:299]: There were a bunch of duplicate fixture around this.

**Comment by marcospri** [comment]
### TODOs folloing our conversation, I'd leave this out for follow up PRs

- [ ] set_focused_user should only happen for instructors

I'll make that explicit in a different PR to allow testing it independently. 

- [x]  ~~set_focused_use seems to be calling H when it could just get the value from our DB~~

  This is unfortunately not possible at the moment. We only store the display name for instructors  (even we do store that name over in H).

- [ ] Move all the canvas js_config specific out to the plugin. Ideally with a nice API to modify the config outside JSConfig


This is worth exploring but probably a big PR own its own.

- [ ] Split the instructor bar configuration in two, editing and grading

Done here.

- [x] Move the grading parts of the grading bar (getting the students) out of JSConfig

File: lms/product/canvas/_plugin/misc.py [hunk 20]
```diff
@@ -8,6 +8,34 @@
 
 
 class CanvasMiscPlugin(MiscPlugin):
+    def post_launch_assignment_hook(self, request, js_config, assignment):
```
Comment (marcospri) [inline lms/product/canvas/_plugin/misc.py:11]: We use a similar `hook` approach for other differences.

The rationale here is that this might no be the perfect abstraction but helps to keep `if is_canvas` or `if is_d2l` out of the main flow. 

Once we have differences around the same thing in different LMSes we try for a better abstraction.

**Comment by robertknight** [comment]
Outside of the changes in this PR, I noticed a duplicated comment in `_show_document` while looking at this:

```py
        # Store the relationship between the assignment and the course
        self.assignment_service.upsert_assignment_membership(
            assignment=assignment,
            user=self.request.user,
            lti_roles=self.request.lti_user.lti_roles,
        )
        # Store the relationship between the assignment and the course
        self.assignment_service.upsert_assignment_groupings(
            assignment, groupings=[self.course]
        )
```

File: lms/resources/_js_config/__init__.py [hunk 21]
```diff
@@ -290,70 +290,32 @@ def add_deep_linking_api(self):
...
-        """
-        Enable the toolbar with controls for instructors in LMS assignments.
+    def enable_toolbar_editing(self):
+        toolbar_config = self._get_toolbar_config()
 
-        :param enable_editing: Whether to enable controls for editing assignment configuration
-        :param enable_grading: Whether to enable grading controls
-        """
+        toolbar_config["editingEnabled"] = True
+        self._config["instructorToolbar"] = toolbar_config
```
Comment (robertknight) [inline lms/resources/_js_config/__init__.py:297]: It would be nice to avoid the duplicated assignment in `enable_toolbar_editing` and `enable_toolbar_grading`. A possible approach would be to change `_get_toolbar_config` to be `_create_toolbar_config` and make that helper responsible for assigning the `instructorToolbar` key.

**Review by robertknight — APPROVED** [review]
The overall picture here makes sense to me. I left a few minor comments and corrections.

**Comment by robertknight** [comment]
Hmm, I'm seeing the instructor toolbar when logged in as `moodlestudent` in the test assignment at https://github.com/hypothesis/lms/pull/5607. There is an "Edit" button, but clicking that shows an error.

Instructor toolbar

**Review by robertknight — CHANGES_REQUESTED** [review]
I encountered an issue with students seeing the instructor toolbar when testing https://github.com/hypothesis/lms/pull/5607 in Moodle and Blackboard, that occurs in this PR but not the main branch. See previous comment.

File: lms/product/product.py [hunk 22]
```diff
@@ -68,6 +68,11 @@ class Product:
...
+    """Wether to use grading in our toolbar."""
```
[the remark this answers is not in the document]
Comment (marcospri) [reply lms/product/product.py:72]: I meant the ram one here

File: tests/unit/lms/product/canvas/_plugin/misc_test.py [hunk 23]
```diff
@@ -1,11 +1,48 @@
...
 
 
 class TestCanvasMiscPlugin:
+    @pytest.mark.parametrize("is_gradable", [True, False])
+    @pytest.mark.parametrize("is_learner", [True, False])
+    @pytest.mark.parametrize("grading_id", [None, sentinel.grading_id])
+    @pytest.mark.parametrize("focused_user", [None, sentinel.focused_user])
+    def test_post_launch_assignment_hook(
+        self,
+        request,
```
[the remark this answers is not in the document]
Comment (marcospri) [reply tests/unit/lms/product/canvas/_plugin/misc_test.py:17]: It's a pytest fixture:

https://docs.pytest.org/en/6.2.x/reference.html#std-fixture-request

We use the `request.getfixturevalue` here to allow a test to be parameterized with fixtures rather  than values.

**Comment by marcospri** [comment]
> Hmm, I'm seeing the instructor toolbar when logged in as `moodlestudent` in the test assignment at #5607. There is an "Edit" button, but clicking that shows an error.
> 
> Instructor toolbar
> Edit: Same in the Blackboard assignment linked in #5607.

Fixed that on this commit:
https://github.com/hypothesis/lms/pull/5623/commits/53a6eca97f16a0ed7db975145bdfb52aac32739f

It was also happening on the D2L example assignment & user I wrote on the testing instructions but I was too focused on the grading side and didn't notice.

**Review by robertknight — APPROVED** [review]
